### PR TITLE
implement per-PWAD savegame directories

### LINKED ISSUE #793: [Feature] Wadname-based savegame dirs
Right now, Woof stores savegames in its root directory. If someone is switching between several addons within a short period of time, it's possible to accidentally overwrite savegames meant for a different PWAD. Therefore, it would be better if there was a way to keep them separate in a more fool-proof way, but also simply in order to maintain some order in the game root directory.

In other ports such as GZDoom, wadname-based savegame directories are created to avoid such situations. It's actually pretty much like the autoload folder works right now, just for savegames.

Suggested realization:
- Introduction of a CVAR to set up directory for savegames (may default to either root dir or e.g. %root%\savegames)
- Port automatically creates subdirs inside of the specified savegame dir with names based on active IWAD/PWAD (e.g. "doom.wad" for Doom 1 or "requiem.wad" for Doom 2 PWAD "Requiem")
- Only savegames from those subdirs are used, depending on which wad is loaded

Things to consider:
- Port needs to recognize if a wad actually contains maps. Savegame dirs should only be created for those, not mere gameplay mods.
- Needs to work with PWADs for Doom 1+2 as well (obviously).

### PR DESCRIPTION
**Pull request by fabiangreffrath**
Fixes #793

### CODE REVIEW

File: src/d_main.c [hunk 1]
```diff
@@ -2141,6 +2199,28 @@ void D_DoomMain(void)
             D_AddFile(myargv[p]);
     }
 
+  if (!M_CheckParm("-save"))
+  {
+    int i;
+    char *wadname = wadfiles[0], *oldsavegame = basesavegame;
+
+    for (i = mainwadfile; i < numwadfiles; i++)
```
Comment (fabiangreffrath) [inline src/d_main.c:2207]: Should we check for the first PWAD that contains a map, or only check the first one?

File: src/m_misc.c [hunk 2]
```diff
@@ -3263,15 +3263,17 @@ void M_LoadDefaults (void)
...
     fclose(f);
   }
   else
   {
-    I_Printf(VB_WARNING, " Warning: Cannot read %s -- using built-in defaults\n",
+    I_Printf(VB_WARNING, " Warning: Cannot read %s -- using built-in defaults",
                          defaultfile);
   }
 
+  I_Printf(VB_INFO, " savegame directory: %s\n", basesavegame);
```
Comment (fabiangreffrath) [inline src/m_misc.c:3275]: Not sure where the best place is to have this reported.

**Comment by fabiangreffrath** [comment]
We will need a migration strategy for savegames in the base directory, but that's subject to a separate PR.

File: src/m_misc.c (hunk 2 above)
Comment (rfomin) [reply src/m_misc.c:3275]: I don't think we should be reporting it at all.

File: src/d_main.c (hunk 1 above)
Comment (rfomin) [reply src/d_main.c:2207]: I like the current approach.

File: src/m_misc.c (hunk 2 above)
Comment (rfomin) [reply src/m_misc.c:3275]: I don't think we should be reporting it at all.

File: src/d_main.c [hunk 3]
```diff
@@ -2141,6 +2199,28 @@ void D_DoomMain(void)
...
+        break;
+      }
+    }
+
+    basesavegame = M_StringJoin(oldsavegame, DIR_SEPARATOR_S,
+                                M_BaseName(wadname), NULL);
+    free(oldsavegame);
+
+    NormalizeSlashes(basesavegame);
+    M_MakeDirectory(basesavegame);
```
Comment (rfomin) [inline src/d_main.c:2221]: I think we should create a `saves/` directory, where we will put the subdirectories of the PWAD saves.

File: src/m_misc.c (hunk 2 above)
Comment (fabiangreffrath) [reply src/m_misc.c:3275]: Hm, okay. 🤔

File: src/d_main.c (hunk 3 above)
Comment (fabiangreffrath) [reply src/d_main.c:2221]: Choco/Crispy use `savegames/`.

Comment (rfomin) [reply src/d_main.c:2221]: I didn't know that, on Windows it saves in the executable directory. Surely `savegames/` is better.

**Review by rfomin — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
Doesn't work with WAD-in-ZIP. 😞

**Comment by rfomin** [comment]
> Doesn't work with WAD-in-ZIP. 😞

Strange, it works on Windows.

**Comment by fabiangreffrath** [comment]
ZIP loading was a red herring! I tried to load `Doom_Sound_Bulb.zip` and the WAD files in there triggered two bugs! 

1. They contain lumps with 8-char names, but lump names are not terminated in `fileinfo[i].name`.
2. They caused an infinite loop in `FileContainsMaps()`, because I didn't have an exit strategy yet in my "goto-preventing" `while()` loop for valid WAD files that don't contain any maps.

**Comment by rfomin** [comment]
> We will need a migration strategy for savegames in the base directory, but that's subject to a separate PR.

I have a autoload directory with many items. I don't believe we can utilize the header info with PWAD names from the save file because it's impossible to determine the main PWAD (the one that has the levels). Thus, the only way to move files automatically is by using an auto-generated save name like `MAP01 (GD)` (we move it to the gd.wad subdirectory). What if the user edited the names of the saved games? Then we won't be able to move the files automatically.

So I think the best way is to always show old saves in the menu, perhaps with a filter based on the currently loaded PWADs.

**Comment by rfomin** [comment]
Another idea is to move old saves to a subdirectory only if the user started the game with an exactly matching IWAD/PWAD list. We show all other old saves that are incompatible with the current session grey out.

**Pull request merged**